index.rst

```diff
@@ -55,78 +55,452 @@
+
+ - accept data ID expressions in fhe form of partial SQL where clauses;
+
+ - construct and execute a SELECT query that inner-joins the relevant data IDs and applies the user's data ID expressions.
+
+This represents a complete redesign of the system of managing metadata in a Data Repository.  It replaces the simple, raw-data-centric registry database and the APIs for interacting it with with a multi-table database that manages all datasets in a repository.  To represent the results of the queries against this database in Python, it also involves a replacing the dictionary-based data ID concept with a more object-oriented system that can hold relationship information.  These interfaces are more naturally a part of the Butler Library than the SuperTask Library, and we expect the design sketch described in this section evolve in the course of future Butler Library design work.  However, we do not expect this evolution to require significant changes to the rest of the SuperTask Library design.
+
+In the new system, the combination of a dictionary-style data ID and a dataset type name becomes an instance of the :py:class:`Dataset` class.  A key-value pair in that dictionary becomes an instance of the :py:class:`Unit` class (for "unit of data"); a :py:class:`Dataset` instance is conceptually a tuple of :py:class:`Units <Unit>`.  A set of :py:class:`Units <Unit>` and py:class:`Datasets <Dataset>` naturally forms a graph-like data structure called a :py:class:`RepoGraph`, which represents (a subset of) a Data Repository.
+
+.. py:class:: Dataset
+
+    A concrete subclass of the abstract base class :py:class:`Dataset` represents a Butler dataset type: a combination of a name, a storage format, path template, and a set of concrete :py:class:`Unit` subclass type objects that define the units of data that label an instance of the dataset.  If, for example, ``Coadd`` is a :py:class:`Dataset` subclass, the corresponding unit classes might be those for :py:class:`Tract`, :py:class:`Patch`, and :py:class:`Filter`.
+
+    An instance of a :py:class:`Dataset` subclass is thus a handle to a particular Butler dataset; it is the only required argument to ``Butler.get`` in the new system, and one of only two required arguments to :py:class:`Butler.put` (the other being the actual object to store).
+
+    :py:class:`Dataset` subclasses are typically created dynamically (usually via a :py:class:DatasetField` that is part of a :py:class:`SuperTask's <SuperTask>` config class).
+
+    .. py:staticmethod:: subclass(name, UnitClasses)
+
+        Define a new :py:class:`Dataset` subclass dyamically with the given name, with instances of the new class required to hold instances of exactly the given :py:class:`Unit` subclasses (via a named attribute for each :py:class:`Unit` subclass).
+
+    .. py:attribute:: units
+
+        A dictionary containing the units that identify this dataset, with unit type names as keys and :py:class:`Unit` instances as values.
```

Comment:
Yes, this was a change spurred by my (aborted) attempt at examples, because there were a few places it led to much more concise code.  I'm still not really attached to either approach, though.